src/hydra-queue-runner/build-remote.cc

```diff
@@ -215,6 +215,9 @@ static void handshake(Machine::Connection & conn, unsigned int repeats)
         throw Error("unsupported ‘nix-store --serve’ protocol version on ‘%1%’", conn.machine->sshName);
     if (GET_PROTOCOL_MINOR(conn.remoteVersion) < 3 && repeats > 0)
         throw Error("machine ‘%1%’ does not support repeating a build; please upgrade it to Nix 1.12", conn.machine->sshName);
+
+    // Do not attempt to speak a newer version of the protocol
+    conn.remoteVersion = std::min(conn.remoteVersion, our_version);
```

Comment:
I am shocked we got away with not having this...ever! If the two sides don't agree on a version, how can anything work?!